test/backup-restore/mzcompose.py

```diff
@@ -0,0 +1,162 @@
+        name="persistcli",
+        config={
+            # TODO: depends!
+            "mzbuild": "jobs"
+        },
+    ),
+]
+
+
+def workflow_default(c: Composition) -> None:
+    # TODO: extract common substrings
+
+    # Enable versioning for the Persist bucket
+    c.up("minio")
+    c.run(
+        "mc",
+        "alias",
+        "set",
+        "persist",
+        "http://minio:9000/",
+        "minioadmin",
+        "minioadmin",
+    )
+    c.run("mc", "version", "enable", "persist/persist")
+    blob_uri = "s3://minioadmin:minioadmin@persist/persist?endpoint=http://minio:9000/&region=minio"
+
+    # Set up a backupable connection for CRDB
+    c.up("cockroach")
+    c.run("mc", "mb", "persist/crdb-backup")
+    c.exec(
+        "cockroach",
+        "cockroach",
+        "sql",
+        "--insecure",
+        "-e",
+        "CREATE EXTERNAL CONNECTION backup_bucket as "
+        "'s3://persist/crdb-backup?AWS_ENDPOINT=http://minio:9000/&AWS_REGION=minio&"
+        "AWS_ACCESS_KEY_ID=minioadmin&AWS_SECRET_ACCESS_KEY=minioadmin';"
+        "SHOW CREATE ALL EXTERNAL CONNECTIONS;",
+    )
+
+    # Start Materialize, and set up some basic state in it
+    c.up("materialized")
+    c.up("testdrive", persistent=True)
+    c.testdrive(
+        dedent(
+            """
+                > DROP TABLE IF EXISTS numbers;
+                > CREATE TABLE IF NOT EXISTS numbers (id BIGINT);
+                > INSERT INTO numbers SELECT * from generate_series(1, 1);
+                > INSERT INTO numbers SELECT * from generate_series(1, 10);
+                > INSERT INTO numbers SELECT * from generate_series(1, 100);
+                """
+        )
+    )
+
+    # Run a manual CRDB backup
+    c.exec(
+        "cockroach",
+        "cockroach",
+        "sql",
+        "--insecure",
+        "-e",
+        "BACKUP INTO 'external://backup_bucket';",
+    )
+    c.exec(
+        "cockroach",
+        "cockroach",
+        "sql",
+        "--insecure",
+        "-e",
+        "SHOW BACKUPS IN 'external://backup_bucket';",
+    )
+
+    # Make further updates to Materialize's state
+    for i in range(0, 100):
+        # TODO: This seems to be enough to produce interesting shard state;
+        # ie. if we remove the restore-blob step we can see the restore fail.
+        # Is there any cheaper or more obvious way to do that?
+        c.testdrive(
+            dedent(
+                """
+                    > INSERT INTO numbers SELECT * from generate_series(1, 1);
+                    > INSERT INTO numbers SELECT * from generate_series(1, 10);
+                    > INSERT INTO numbers SELECT * from generate_series(1, 100);
+                    """
+            )
+        )
+
+    # Stop the running cluster and execute a restore
+    c.kill(
+        "materialized"
+    )  # Very important that MZ is not running during the restore process!
+    c.exec(
+        "cockroach", "cockroach", "sql", "--insecure", "-e", "DROP DATABASE defaultdb;"
```

Comment:
This immediately puts the database in an invalid state. It does not return to a valid state until the successful `restore-blob` call below. This is sort of a "critical section"; missing one of these steps, or having a materialize instance running while either of these things is happening, is likely to cause crashes or other breakage... but that breakage is expected.